### PR TITLE
Add fuse-conv-bn method for Conv2d

### PR DESCRIPTION
**Pull request by jamjamjon**
### 1. Add a fuse-conv-bn method: `absorb_bn(&self, bn: &BatchNorm) -> Result<Self>` for Conv2d.
The BatchNorm2d layer will be absorb into Conv2d, and return a Conv2d which has bias parameter. Whether the Conv2d layer has bias or not, this method can fuse adjacent Conv2d and BatchNorm2d layers together.

**usage:**
```
...
let bn = batch_norm(c2, 1e-3, vb.pp("bn"))?;
let conv = conv2d_no_bias(c1, c2, k, cfg, vb.pp("conv"))?.absorb_bn(&bn)?;
...
```

### 2. Expose another filed from BatchNorm in order to fuse Conv2d and BatchNorm.
```
impl BatchNorm {
    ...
    pub fn eps(&self) -> f64 {
        self.eps
    } 
}
```

### 3. A slight modification to yolo-v8 example: 

- Removed `bn` field in `ConvBlock`
- Add fused_conv for inference

### Env
- GPU: NVIDIA GeForce RTX 3060, 12036MiB)
- CPU: 12th Gen Intel(R) Core(TM) i5-12400F
- OS: Ubuntu 23.04
- Model: yolov8s.pt, yolov8s.onnx, yolov8s.safetensors

### Example
`cargo run --release --features cuda  --example yolo-v8 -- --model ./yolov8s.safetensors --which s candle-examples/examples/yolo-v8/assets/bike.jpg`

###  20 times inference of `let predictions = model.forward(&image_t)?.squeeze(0)?;`

**Before** ==>  `Silu(BN(Conv2d(xs)))`  
Forward : 61.892098ms
Forward : 29.135077ms
Forward : 27.852497ms
Forward : 28.609387ms
Forward : 27.01425ms
Forward : 27.019482ms
Forward : 26.935323ms
Forward : 27.050945ms
Forward : 26.719977ms
Forward : 26.942656ms
Forward : 25.886604ms
Forward : 27.8262ms
Forward : 26.914334ms
Forward : 27.045991ms
Forward : 26.954339ms
Forward : 26.752925ms
Forward : 26.563086ms
Forward : 26.795354ms
Forward : 26.647943ms
Forward : 26.849186ms
Forward : 26.854436ms

**After** ==>  `Silu(Conv2d(xs))`  
Forward : 41.397483ms
Forward : 24.402086ms
Forward : 24.167527ms
Forward : 23.946848ms
Forward : 22.95527ms
Forward : 22.527519ms
Forward : 22.992553ms
Forward : 22.022945ms
Forward : 21.245696ms
Forward : 22.592846ms
Forward : 23.00923ms
Forward : 22.073522ms
Forward : 21.979598ms
Forward : 23.03203ms
Forward : 22.92703ms
Forward : 21.764525ms
Forward : 22.368902ms
Forward : 23.05773ms
Forward : 22.122301ms
Forward : 20.983996ms
Forward : 22.300197ms

**Accelerated by ~4ms**

### CODE REVIEW

File: candle-nn/src/conv.rs [hunk 1]
```diff
@@ -115,6 +116,21 @@ impl Conv2d {
     pub fn bias(&self) -> Option<&Tensor> {
         self.bias.as_ref()
     }
+
+    pub fn absorb_bn(&self, bn: &BatchNorm) -> Result<Self> {
+        let (w_bn, b_bn) = bn.weight_and_bias().unwrap();
```
Comment (LaurentMazare) [inline candle-nn/src/conv.rs:121]: This should be an error rather than a unwrap, just use `candle::bail!("batch norm does not have weight_and_bias")` or something similar.

Comment (jamjamjon) [reply candle-nn/src/conv.rs]: Fixed.

**Comment by LaurentMazare** [comment]
Please run rustfmt and clippy (see the CI errors).

**Comment by LaurentMazare** [comment]
Looks pretty good. Could you confirm that you checked that the yolo model outputs were unchanged after applying this? (we don't have a test covering this model yet)

**Comment by jamjamjon** [comment]
Here is the output:

**Not fused**
processing candle-examples/examples/yolo-v8/assets/bike.jpg
generated predictions Tensor[dims 84, 5460; f32, cuda:0]
person: Bbox { xmin: 86.70215, ymin: 72.80286, xmax: 197.06848, ymax: 345.34613, confidence: 0.85669655, data: [] }
person: Bbox { xmin: 213.19174, ymin: 65.66339, xmax: 326.40512, ymax: 360.43457, confidence: 0.8543925, data: [] }
person: Bbox { xmin: 378.09106, ymin: 84.67178, xmax: 480.59656, ymax: 346.74576, confidence: 0.82226956, data: [] }
person: Bbox { xmin: 323.472, ymin: 64.844574, xmax: 412.26825, ymax: 304.65915, confidence: 0.8124224, data: [] }
person: Bbox { xmin: 13.493347, ymin: 62.32115, xmax: 53.200027, ymax: 192.76765, confidence: 0.7586017, data: [] }
person: Bbox { xmin: 563.3775, ymin: 61.85759, xmax: 621.75116, ymax: 162.48303, confidence: 0.64022607, data: [] }
person: Bbox { xmin: 392.88464, ymin: 13.64006, xmax: 438.71613, ymax: 92.1143, confidence: 0.5986935, data: [] }
person: Bbox { xmin: 387.90167, ymin: 58.17116, xmax: 426.3554, ymax: 120.541, confidence: 0.5346883, data: [] }
person: Bbox { xmin: 0.0001411438, ymin: 54.14986, xmax: 17.51616, ymax: 181.92192, confidence: 0.45644093, data: [] }
person: Bbox { xmin: 491.9308, ymin: 57.25844, xmax: 575.2428, ymax: 163.41861, confidence: 0.40927175, data: [] }
person: Bbox { xmin: 78.410645, ymin: 62.86883, xmax: 106.18094, ymax: 107.6958, confidence: 0.37573242, data: [] }
person: Bbox { xmin: 61.32128, ymin: 70.5005, xmax: 88.35351, ymax: 154.13112, confidence: 0.36326504, data: [] }
person: Bbox { xmin: 561.8749, ymin: 62.875626, xmax: 613.89124, ymax: 104.10129, confidence: 0.33813742, data: [] }
person: Bbox { xmin: 169.88559, ymin: 62.11824, xmax: 206.20975, ymax: 130.0948, confidence: 0.31063026, data: [] }
person: Bbox { xmin: 625.9895, ymin: 49.59937, xmax: 638.81323, ymax: 84.50955, confidence: 0.30136093, data: [] }
bicycle: Bbox { xmin: 221.59367, ymin: 196.18845, xmax: 317.95697, ymax: 394.34482, confidence: 0.86914253, data: [] }
bicycle: Bbox { xmin: 91.21129, ymin: 194.49448, xmax: 185.03185, ymax: 376.82565, confidence: 0.78296405, data: [] }
bicycle: Bbox { xmin: 330.97885, ymin: 187.6524, xmax: 393.27243, ymax: 351.8576, confidence: 0.78266, data: [] }
bicycle: Bbox { xmin: 385.92184, ymin: 202.51932, xmax: 461.12454, ymax: 383.60672, confidence: 0.76453125, data: [] }
bicycle: Bbox { xmin: 184.69647, ymin: 178.77426, xmax: 204.20868, ymax: 241.60062, confidence: 0.25113082, data: [] }
car: Bbox { xmin: 283.788, ymin: 63.868454, xmax: 356.5054, ymax: 136.17131, confidence: 0.32286114, data: [] }
motorbike: Bbox { xmin: 491.94183, ymin: 99.806946, xmax: 568.4294, ymax: 215.28107, confidence: 0.8369786, data: [] }
motorbike: Bbox { xmin: 562.19165, ymin: 84.61237, xmax: 622.74646, ymax: 176.6341, confidence: 0.6846957, data: [] }
dog: Bbox { xmin: 57.65777, ymin: 148.95132, xmax: 90.12031, ymax: 193.92679, confidence: 0.83050436, data: [] }
writing "candle-examples/examples/yolo-v8/assets/bike.pp.jpg"


**Fused**
processing candle-examples/examples/yolo-v8/assets/bike.jpg
generated predictions Tensor[dims 84, 5460; f32, cuda:0]
person: Bbox { xmin: 86.70215, ymin: 72.80289, xmax: 197.06848, ymax: 345.34613, confidence: 0.85669655, data: [] }
person: Bbox { xmin: 213.19174, ymin: 65.66339, xmax: 326.40512, ymax: 360.43457, confidence: 0.85439265, data: [] }
person: Bbox { xmin: 378.09106, ymin: 84.67178, xmax: 480.59656, ymax: 346.74576, confidence: 0.82226986, data: [] }
person: Bbox { xmin: 323.472, ymin: 64.844574, xmax: 412.26825, ymax: 304.65912, confidence: 0.8124227, data: [] }
person: Bbox { xmin: 13.493349, ymin: 62.32115, xmax: 53.200035, ymax: 192.76767, confidence: 0.7586018, data: [] }
person: Bbox { xmin: 563.3775, ymin: 61.85758, xmax: 621.75116, ymax: 162.48303, confidence: 0.64022607, data: [] }
person: Bbox { xmin: 392.88464, ymin: 13.640068, xmax: 438.7162, ymax: 92.1143, confidence: 0.5986935, data: [] }
person: Bbox { xmin: 387.90167, ymin: 58.17117, xmax: 426.3554, ymax: 120.541016, confidence: 0.5346887, data: [] }
person: Bbox { xmin: 0.0001373291, ymin: 54.14986, xmax: 17.51616, ymax: 181.92192, confidence: 0.45644078, data: [] }
person: Bbox { xmin: 491.9308, ymin: 57.25844, xmax: 575.2428, ymax: 163.41861, confidence: 0.40927154, data: [] }
person: Bbox { xmin: 78.41065, ymin: 62.86883, xmax: 106.18093, ymax: 107.6958, confidence: 0.3757325, data: [] }
person: Bbox { xmin: 61.32128, ymin: 70.50052, xmax: 88.35351, ymax: 154.13113, confidence: 0.3632654, data: [] }
person: Bbox { xmin: 561.8749, ymin: 62.875633, xmax: 613.89124, ymax: 104.101295, confidence: 0.33813733, data: [] }
person: Bbox { xmin: 169.88559, ymin: 62.11824, xmax: 206.20975, ymax: 130.09482, confidence: 0.31063077, data: [] }
person: Bbox { xmin: 625.9895, ymin: 49.59937, xmax: 638.81323, ymax: 84.50955, confidence: 0.30136073, data: [] }
bicycle: Bbox { xmin: 221.59366, ymin: 196.18845, xmax: 317.957, ymax: 394.34482, confidence: 0.86914235, data: [] }
bicycle: Bbox { xmin: 91.21128, ymin: 194.49445, xmax: 185.03183, ymax: 376.82568, confidence: 0.78296405, data: [] }
bicycle: Bbox { xmin: 330.97888, ymin: 187.6524, xmax: 393.27246, ymax: 351.8576, confidence: 0.7826597, data: [] }
bicycle: Bbox { xmin: 385.92188, ymin: 202.51932, xmax: 461.1245, ymax: 383.60672, confidence: 0.76453143, data: [] }
bicycle: Bbox { xmin: 184.69647, ymin: 178.77426, xmax: 204.20865, ymax: 241.60062, confidence: 0.251131, data: [] }
car: Bbox { xmin: 283.788, ymin: 63.868454, xmax: 356.5054, ymax: 136.1713, confidence: 0.32286137, data: [] }
motorbike: Bbox { xmin: 491.94183, ymin: 99.80695, xmax: 568.4294, ymax: 215.28107, confidence: 0.8369787, data: [] }
motorbike: Bbox { xmin: 562.19165, ymin: 84.6124, xmax: 622.74646, ymax: 176.6341, confidence: 0.68469536, data: [] }
dog: Bbox { xmin: 57.657772, ymin: 148.95132, xmax: 90.1203, ymax: 193.92679, confidence: 0.8305042, data: [] }
writing "candle-examples/examples/yolo-v8/assets/bike.pp.jpg"

**Comment by LaurentMazare** [comment]
Ok looks close enough, thanks!
